File: client/src/main/java/ch/ergon/modern/java/client/formula/FormulaEvaluator.java
package ch.ergon.modern.java.client.formula;

import ch.ergon.modern.java.api.formula.BinaryFormula;
import ch.ergon.modern.java.api.formula.Constant;
import ch.ergon.modern.java.api.formula.Formula;
import ch.ergon.modern.java.api.formula.UnaryFormula;

public class FormulaEvaluator {

    public double evaluate(Formula formula) {
        return switch (formula) {
            case Constant c -> c.value();
            case BinaryFormula b -> evaluateBinaryFormula(b);
            case UnaryFormula u -> evaluateUnaryFormula(u);
        };
    }

    private double evaluateBinaryFormula(BinaryFormula formula) {
        return switch (formula.operator()) {
            case ADD -> evaluate(formula.leftSide()) + evaluate(formula.rightSide());
            case SUBTRACT -> evaluate(formula.leftSide()) - evaluate(formula.rightSide());
            case MULTIPLY -> evaluate(formula.leftSide()) * evaluate(formula.rightSide());
            case DIVIDE -> evaluate(formula.leftSide()) / evaluate(formula.rightSide());
        };
    }

    private double evaluateUnaryFormula(UnaryFormula formula) {
        return switch (formula.operator()) {
            case PLUS -> evaluate(formula.formula());
            case MINUS -> -evaluate(formula.formula());
        };
    }
}

File: client/src/main/kotlin/ch/ergon/modern/java/client/formula/kotlin/FormulaEvaluator.kt
package ch.ergon.modern.java.client.formula.kotlin

import ch.ergon.modern.java.api.formula.kotlin.*

fun Formula.evaluate(variables: Map<String, Double>): Double = when (this) {
    is Constant -> value
    is BinaryFormula -> evaluate()
    is UnaryFormula -> evaluate()
    is Variable -> variables[name] ?: throw IllegalStateException("Undefined value for variable '$name'")
}

fun Formula.evaluate(): Double = evaluate(emptyMap())

private fun BinaryFormula.evaluate(): Double = when (operator) {
    BinaryOperator.ADD -> leftSide.evaluate() + rightSide.evaluate()
    BinaryOperator.SUBTRACT -> leftSide.evaluate() - rightSide.evaluate()
    BinaryOperator.MULTIPLY -> leftSide.evaluate() * rightSide.evaluate()
    BinaryOperator.DIVIDE -> leftSide.evaluate() / rightSide.evaluate()
}

private fun UnaryFormula.evaluate(): Double = when (operator) {
    UnaryOperator.PLUS -> formula.evaluate()
    UnaryOperator.MINUS -> -formula.evaluate()
}